cmd/metal-api/internal/headscale/client.go

```diff
@@ -0,0 +1,238 @@
+package headscale
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	headscalecore "github.com/juanfont/headscale"
+	headscalev1 "github.com/juanfont/headscale/gen/go/headscale/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+var (
+	apiKeyExpiration = 24 * 90 * time.Hour
+)
+
+type HeadscaleClient struct {
+	client headscalev1.HeadscaleServiceClient
+
+	address             string
+	ControlPlaneAddress string
+
+	apiKeyPrefix    string
+	oldAPIKeyPrefix string
+
+	mu         sync.Mutex
+	wg         sync.WaitGroup
+	ctx        context.Context
+	conn       *grpc.ClientConn
+	cancelFunc context.CancelFunc
+	logger     *zap.SugaredLogger
+}
+
+func NewHeadscaleClient(addr, controlPlaneAddr, apiKey string, logger *zap.SugaredLogger) (client *HeadscaleClient, err error) {
+	if addr != "" || apiKey != "" {
+		if addr == "" {
+			return nil, fmt.Errorf("headscale address should be set with api key")
+		}
+		if apiKey == "" {
+			return nil, fmt.Errorf("headscale api key should be set with address")
+		}
+	} else {
+		return
+	}
+
+	h := &HeadscaleClient{
+		address:             addr,
+		ControlPlaneAddress: controlPlaneAddr,
+
+		mu:     sync.Mutex{},
+		wg:     sync.WaitGroup{},
+		logger: logger,
+	}
+	h.ctx, h.cancelFunc = context.WithCancel(context.Background())
+
+	// Validate API Key
+	if h.apiKeyPrefix, err = getApiKeyPrefix(apiKey); err != nil {
+		return nil, err
+	}
+
+	if err = h.connect(apiKey); err != nil {
+		return nil, fmt.Errorf("failed to connect to Headscale server: %w", err)
+	}
+
+	if err = h.replaceApiKey(); err != nil {
+		return nil, fmt.Errorf("failed to replace Headscale API Key: %w", err)
+	}
+
+	go h.rotateApiKey()
+
+	return h, nil
+}
+
+// Connect or reconnect to Headscale server
+func (h *HeadscaleClient) connect(apiKey string) (err error) {
+	ctx, cancel := context.WithTimeout(h.ctx, 5*time.Second)
+	defer cancel()
+
+	grpcOptions := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithPerRPCCredentials(tokenAuth{
+			token: apiKey,
+		}),
+	}
+
+	h.conn, err = grpc.DialContext(ctx, h.address, grpcOptions...)
+	if err != nil {
+		return fmt.Errorf("failed to connect to headscale server %s: %w", h.address, err)
+	}
+
+	h.client = headscalev1.NewHeadscaleServiceClient(h.conn)
+
+	return
+}
+
+// Refresh API key when apiKeyExpiration reaches it's half life
+func (h *HeadscaleClient) rotateApiKey() {
```

Comment:
Check if we really need the key rotation as discussed.

Reply:
The key rotation is a little bit more complex than i realized before. First of all, yes we absolutely need it as expired key doesn't work completely. Another concern is that the initial key will be stored in a `Secret`, so when pod restarted it will always try to use key from this secret even if it's expired.

I think the ideal solution would be to rotate that Secret. So the key rotation task would be outside of the metal-api scope. WDYT?